### PR TITLE
Fix: medication X button passes wrong ID and removed flawed archival mechanism that was only used to archive re-staged medications after re-editing

### LINKED ISSUE #2339: [Bug]: Medication X button passes wrong ID to server, unmasking deleteOnCloseRxBox archival bug
### Description

> **Note:** These bug numbers are based on a google document listing medication issues found, not other issues tracked in GitHub.
                                                                                                                                                                                                                                            
  ## Summary
  Two related bugs in the prescription staging UI. Bug `#5` is currently masking Bug `#6` — fixing `#5` without `#6` causes unintended drug archival. Both must ship together.                                                              
                                                                                                                                                                                                                                            
  ### Bug `#5`: X Button Passes Wrong ID to Server [REGRESSION]                                                                                                                                                                             
                                                                                                                                                                                                                                            
  `deletePrescribingDrugFromUI(uiRefId, drugId)` passes `drugId` (DrugReferenceId) to `deletePrescribe()` instead of `uiRefId` (randomId). The server looks up by randomId, finds nothing, and the stash item is never removed.             
                  
  For brand-new drugs, DrugReferenceId defaults to 0 and randomId is a large random number — they will never match. The X button always fails to remove new drugs from the stash.                                                           
                  
  **File:** `SearchDrug3.jsp` — `deletePrescribingDrugFromUI()      

---                                                                                                                                                                     
                                                                                                                                                                                                                                            
  ### Bug `#6`: deleteOnCloseRxBox Flag Never Reset [PRE-EXISTING] [MASKED BY BUG `#5`]                                                                                                                                                             
                                                                                                                                                                                                                                            
  ▎ Note: This bug is currently masked by Bug `#5`. Because Bug `#5` passes the wrong parameter to deletePrescribe(), the server-side archival lookup fails silently. If Bug `#5` is fixed without also fixing Bug `#6`, the issue described below  
  becomes active.                                                                                                                                                                                                                           
                                                                                                                                                                                                                                            
When "Edit Rx" is clicked, updateDeleteOnCloseRxBox() sets deleteOnCloseRxBox to 'true'. It is never reset to 'false'. After Edit Rx, every subsequent X button click also triggers deleteRxOnCloseRxBox(), which archives the drug from the database without confirmation.
                                                                                                                                                                                                                                            
  File: SearchDrug3.jsp — deleteOnCloseRxBox hidden input, updateDeleteOnCloseRxBox()                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                            
  ### Proposed Fix                                                                                                                                                                                                                              
                                                                                                                                                                                                                                            
  - Bug `#5`: Change this.deletePrescribe(drugId) to this.deletePrescribe(uiRefId) in deletePrescribingDrugFromUI()                                                                                                                           
  - Bug `#6`: Reset deleteOnCloseRxBox to 'false' after the lightwindow closes or after Save & Print completes
                                                                                                                                                                                                                                            
  ---                                                                  

### Steps to Reproduce

### Bug `#5`  

**Steps to reproduce:**                                                                                                                                                                                                                   
  1. Open the patient's medication page
  2. Add any drug (not through ReRx — through the search bar only)
  3. Click the X button on the drug card to remove it                                                                                                                                                                                       
  4. Check stash count via browser console:
     ```js                                                                                                                                                                                                                                  
     new Ajax.Request(ctx + '/oscarRx/WriteScript.do?parameterValue=checkNoStashItem&rand=1', {method:'get', onSuccess:function(t){console.log('stash items: ' + t.responseText.evalJSON().NoStashItem)}})                                                                                                                                                                    

### Bug `#6`

 **Steps to reproduce (with Bug `#5` fix applied):**                                                                                                                                                                                         
  1. Add Drug A and Drug B, click "Save & Print" — both print and are saved to DB
  2. Click "Edit Rx" — this sets deleteOnCloseRxBox = 'true'                                                                                                                                                                                
  3. Click X on Drug B's staging card                                                                                                                                                                    

---

### Expected Result

### Bug `#5`
**Expected:** stash items: 0 — drug removed from UI and server stash        

### Bug `#6`
**Expected:** Drug B removed from staging only (the X button should un-stage, not archive)   

---

### Actual Result

### Bug `#5`
**Actual:** stash items: 1 — drug removed from UI only 

### Bug `#6`
**Actual:** Because deleteOnCloseRxBox is still 'true', deletePrescribe() also calls deleteRxOnCloseRxBox(), which archives Drug B from the DB — the same effect as clicking "Del" in the drug profile, but without a confirmation dialog. Drug B shows crossed out and disappears from the default "Current" medication history view after refreshing

---

### Module/Area

Other

### Error Messages/Logs

```shell
N/A
```

### Version/Branch

develop

### Checklist

- [x] I have searched for similar issues
- [x] I can reproduce this consistently

### LINKED ISSUE #2341: [Bug]: Missing response content-type on DeleteRxOnCloseRxBox causes Firefox XML parsing error
### Description

> **Note:** These bug numbers are based on a google document listing medication issues found, not other issues tracked in GitHub.                                                                                                         
                                                                                                                                                                                                                                            
  ## Summary                                                                                                                                                                                                                                
  `RxDeleteRx2Action.DeleteRxOnCloseRxBox()` writes JSON to the response without setting the content type. Firefox attempts to parse it as XML and logs a parsing error.                                                                    
                                                                                                                                                                                                                                            
  ### Bug `#7`: Missing Content-Type on JSON Response [PRE-EXISTING]                                                                                                                                                                        
                                                                                                                                                                                                                                            
  **File:** `RxDeleteRx2Action.java` — `DeleteRxOnCloseRxBox()`                                                                                                                                                                             
                                                                           
                                                                                                                                                                                                                                            
  ## Proposed Fix                                                                                                                                                                                                                           
  Add `response.setContentType("application/json")` before writing the JSON response body. 

---

### Steps to Reproduce

**Steps to reproduce:**                                                                                                                                                                                                                   
  1. Open the patient's medication page in Firefox
  2. Add a drug, Save & Print, then click Edit Rx                                                                                                                                                                                           
  3. Click the X button on a staged drug
  4. Open Firefox DevTools → Console                                                                                                                                                                                                                                                                                                                                                                                                            

---

### Expected Result

**Expected:** Clean AJAX response  

---

### Actual Result

 **Actual:** "XML Parsing Error: not well-formed" for the `deleteRx.do?parameterValue=DeleteRxOnCloseRxBox` request. Cosmetic — archival still works, but may mask real errors 

---

### Module/Area

Other

### Error Messages/Logs

```shell

```

### Version/Branch

develop

### Checklist

- [x] I have searched for similar issues
- [x] I can reproduce this consistently

### PR DESCRIPTION
**Pull request by LiamStanziani**
## Summary
  Fixes two interdependent bugs in the prescription staging UI where the X button on drug cards failed to remove drugs from the server stash, and a hidden flag caused unintended permanent archival of medications from the database.

  Fixes #2339
  Fixes #2341 

> **Note:** These bug numbers are based on a google document listing medication issues found, not other issues tracked in GitHub.

  ## Problem
  **Bug `#5` (Regression):** `deletePrescribingDrugFromUI()` passed `drugId`
  (DrugReferenceId) to `deletePrescribe()` instead of `uiRefId` (randomId).
  The server looks up stash items by randomId, so the lookup always failed silently
  — the X button removed the drug from the UI but left it in the server stash.

  **Bug `#6` (Pre-existing, masked by Bug `#5`):** The `deleteOnCloseRxBox` hidden
  field was set to `'true'` when "Edit Rx" was clicked but never reset. This caused
  `deletePrescribe()` to also call `deleteRxOnCloseRxBox()`, which permanently
  archived drugs from the database without user confirmation. Because Bug `#5`
  prevented `deletePrescribe()` from reaching that code path, this bug was masked.

  ## Solution
  **Bug `#5`:** Pass `uiRefId` (the randomId) instead of `drugId` to
  `deletePrescribe()`. Remove the now-unused `drugId` parameter from
  `deletePrescribingDrugFromUI()`.

**Bug `#6`:** Remove the entire `deleteOnCloseRxBox` mechanism — a flawed 2010-era
  feature (commit 75940c0, SourceForge Bug: https://sourceforge.net/p/oscarmcmaster/bugs/1336/) that silently changed the X button from
  "remove from staging" to "permanently archive from database" based on hidden
  state, with no visual indication or confirmation. This is unsafe behavior for a
  healthcare system. Removed: the hidden input, flag setter, conditional check in
  `deletePrescribe()`, the `deleteRxOnCloseRxBox()` function, and the
  server-side `DeleteRxOnCloseRxBox()` method in `RxDeleteRx2Action.java`.

**Bug `#7`:** Removed the underlying functionality, this is no longer applicable, resolved in this PR

  ## Additional improvements
  - **CSRF protection:** `deletePrescribe()` changed from GET to POST with CSRF
    token header, so it is now covered by CSRFGuard's protected methods
    (`POST,PUT,DELETE`)
  - **Prototype.js removal:** Replaced `Ajax.Request` with `fetch` and
    Prototype `$()` calls with `document.getElementById()`
  - **jQuery removal:** Replaced `jQuery("#...").remove()` with vanilla
    `document.getElementById()?.remove()`
  - **Error handling:** Added error logging for failed stash removal requests
  - **JSDoc fix:** Corrected misleading `@param` documentation on
    `uncheckReRxForExistingPrescribedDrug()`
- **Additional javadoc fixes and additions**

## Summary by Sourcery

Fix prescription staging UI deletion so it targets the correct server-side item and remove the legacy hidden-flag mechanism that could silently archive medications.

Bug Fixes:
- Ensure the prescription card X button deletes the correct staged medication on the server by using the UI reference ID instead of the drug ID.
- Eliminate the deleteOnCloseRxBox-based archival flow, including its hidden field, updater, and delete-on-close handler, to prevent unintended permanent archival of medications.

Enhancements:
- Simplify the prescription deletion codepath by removing unused parameters and dead archival-related logic from the staging UI.

---
## Summary by cubic
Fixes the prescription staging X button so it deletes the correct staged item on the server and removes the legacy archival path that could silently archive medications. Also hardens deletion with CSRF-protected POST, safer logging, and better server-side handling.

- **Bug Fixes**
  - Pass `uiRefId` to `deletePrescribe()` instead of `drugId`; remove `drugId` from `deletePrescribingDrugFromUI()` and update the caller.
  - Remove the `deleteOnCloseRxBox` mechanism across client and server, including the client function and the server `DeleteRxOnCloseRxBox` action.

- **Refactors**
  - Switch stash deletion to `fetch` with CSRF and `X-Requested-With` header.
  - Add a null/empty check for `drugList` in bulk deletion and warn instead of proceeding; update Javadoc/JSDoc and add a header comment to `SearchDrug3.jsp`.
  - Encode logged method parameters with `org.owasp.encoder.Encode` to avoid unsafe log output.

<sup>Written for commit 71a7a86e8f6c00f8a7c6b279776c1c5cd6d4726b. Summary will update on new commits.</sup>

## Summary by Sourcery

Fix prescription staging deletion to target the correct staged medication and remove the legacy delete-on-close archival path, while modernizing the client-side request and security for stash deletion.

Bug Fixes:
- Ensure the X button on prescription cards deletes the correct staged medication on the server by using the UI reference ID.
- Remove the deleteOnCloseRxBox-based delete-on-close mechanism that could silently archive prescriptions.

Enhancements:
- Switch prescription stash deletion to a CSRF-protected POST request using fetch and vanilla DOM APIs.
- Simplify prescription deletion logic by removing unused parameters, dead archival code paths, and correcting related JSDoc.

## Summary by CodeRabbit

* **Improvements**
  * Switched deletion requests to a safer POST flow with CSRF protection and better HTTP error handling.
  * Simplified UI deletion logic so items remove reliably using a single stable identifier.
  * Safer DOM updates with helper routines to prevent UI errors.

* **Bug Fixes**
  * Added handling for bulk-deletion paths and a guard for missing input to avoid failed deletions and log warnings.

### CODE REVIEW

**Comment by LiamStanziani** [comment]
@sourceryai review

**Comment by LiamStanziani** [comment]
@coderabbitai review

File: src/main/webapp/oscarRx/SearchDrug3.jsp [hunk 1]
```diff
@@ -1728,42 +1727,13 @@
     new Ajax.Request(url, {
       method: 'get', parameters: data, onSuccess: function (transport) {
         // updateCurrentInteractions();
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/SearchDrug3.jsp]: Will be fixed in the next commit

File: src/main/webapp/oscarRx/SearchDrug3.jsp [hunk 2]
```diff
@@ -1723,47 +1722,26 @@
...
+    headers['<csrf:tokenname/>'] = document.querySelector('#drugForm input[name="<csrf:tokenname/>"]').value;
 
-        jQuery("#set_" + randomId).remove();
-        jQuery("#prescriptionMoreLessLink_" + randomId).remove();
-        jQuery("#deleteMedicationFromPrescription_" + randomId).remove();
+    fetch(url, {
+      method: 'POST',
+      headers: headers,
+      body: new URLSearchParams({parameterValue: 'deletePrescribe', randomId: randomId})
+    }).then(function (response) {
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/SearchDrug3.jsp:1754]: Will be fixed in the next commit

[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/SearchDrug3.jsp:1754]: Will be fixed in the next commit

File: src/main/webapp/oscarRx/SearchDrug3.jsp [hunk 3]
```diff
@@ -2641,18 +2614,17 @@
...
      * Deletes a prescribing drug from UI and calls deletePrescribe.
      * @param uiRefId The unique id for referencing the UI element.
-     * @param drugId The id of the drug to delete.
      */
-    function deletePrescribingDrugFromUI(uiRefId, drugId) {
+    function deletePrescribingDrugFromUI(uiRefId) {
       this.removeElementFromUI(this.getPrescribingDrugCardByUiRefId(uiRefId));
-      this.deletePrescribe(drugId);
+      this.deletePrescribe(uiRefId);
     }
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/SearchDrug3.jsp:2649]: The UI actually already removes the care before the deletePrescribe is called, so I don't see any issue with still setting this

File: src/main/webapp/oscarRx/SearchDrug3.jsp [hunk 4]
```diff
@@ -1015,7 +1015,6 @@
                         <div id="rxText"></div>
                         <%-- Prescriptions are staged here via the prescribe.jsp widget --%>
 
-                        <input type="hidden" id="deleteOnCloseRxBox" value="false"/>
                         <input type="hidden" property="demographicNo" value="<%=patient.getDemographicNo()%>"/>
 
                       </div>
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/webapp/oscarRx/SearchDrug3.jsp:1041]: Will be fixed in the next commit

File: src/main/java/ca/openosp/openo/prescript/pageUtil/RxDeleteRx2Action.java [hunk 5]
```diff
@@ -135,6 +133,10 @@ public String execute()
             response.sendRedirect("error.html");
             return null;
         }
+        if (drugList == null || drugList.isBlank()) {
+            MiscUtils.getLogger().warn("RxDeleteRx2Action: no drugList provided for bulk deletion (parameterValue=" + method + ")");
```
[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/prescript/pageUtil/RxDeleteRx2Action.java]: Will be fixed in the next commit

[the remark this answers is not in the document]
Comment (LiamStanziani) [reply src/main/java/ca/openosp/openo/prescript/pageUtil/RxDeleteRx2Action.java]: Will be fixed in the next commit (with coderabbit's approach)

**Comment by LiamStanziani** [comment]
### Context on the `deleteOnCloseRxBox` removal

  The `deleteOnCloseRxBox` mechanism was added in March 2010 to fix [SourceForge Bug #1336](https://sourceforge.net/p/oscarmcmaster/bugs/1336/) (commit `75940c0`, Bug ID 2968933). The report stated that after clicking "Save and Prescribe" (the old name for "Save and Print", renamed in commit `b2fc152`), editing the prescription, and removing a drug, the drug would still appear after re-entering the module.

  However, the same developer (`ygmgbg`) had added save-before-preview behavior 4 months earlier in commit
  `2a73cc3058`:

  > *"save drugs before show up lightwindow preview"*

  This meant drugs were already committed to the database **before** the print preview opened and before the user could click "Edit Rx." Removing a drug during editing only removed it from the session stash, the previously saved database record was untouched. The `deleteOnCloseRxBox` mechanism was a workaround that silently archived drugs from the database when the X button was clicked during editing, rather than addressing the underlying save flow. To connect this, Bug `#4` in the google document explores whether "Save And Print" should defer the database save until after printing/faxing. If that approach is adopted (Carmen's proposal), it would naturally resolve the original SourceForge reporter's concern without needing a mechanism like deleteOnCloseRxBox. If we go with Keith's approach, the "Edit Rx" click would archive the entire script that was just saved, returning all drugs to stash-only state, rather than the old per-drug archival on individual X clicks. In my suggested approach (save the medications but clear the stash on "Save and Print"), this functionality would also be unnecessary, since there's nothing left in the stash to edit, the X button archival mechanism has no purpose. This holds true for all three proposed approaches.


  This mechanism was problematic because:
  - The X button silently changed behavior from "remove from staging" to "permanently archive from database" based
  on hidden state
  - There was no visual indication or confirmation for the destructive action
  - The flag was never reset, so the archival behavior leaked beyond the edit session
  - And as explained above, Bug `#4` resolution will most likely make this functionality dead code

  I have removed the entire mechanism (client-side and server-side `DeleteRxOnCloseRxBox`) as it is potentially unsafe for a healthcare system. The X button now consistently only removes drugs from the session stash.

**Review by D3V41 — APPROVED** [review]
